python/pyarrow/_flight.pyx

```diff
@@ -1707,16 +1735,26 @@ cdef class ClientAuthSender(_Weakrefable):
         return result
 
 
-cdef CStatus _data_stream_next(void* self, CFlightPayload* payload) except *:
+cdef CStatus _data_stream_next(void* c_state,
+                               CFlightPayload* payload) except *:
     """Callback for implementing FlightDataStream in Python."""
+    server_state = <object> c_state
+    try:
+        if not isinstance(server_state.this, GeneratorStream):
+            raise RuntimeError(
+                "self object in callback is not GeneratorStream")
+        stream = <GeneratorStream> server_state.this
+        return _data_stream_do_next(stream, server_state, payload)
+    except Exception as e:
+        server_state.log_exception("do_get", e)
```

Comment:
Is this a "do_get" operation?
[In `_do_get`](https://github.com/apache/arrow/pull/12659/files#diff-0ed358f5d42920d7f94cc500791976a2c158c4d72f4a6b231393534b2d13683bR1957), name is also "do_get".

Reply:
Yes, basically DoGet returns a Stream and the Stream is consumed by the C++ side to finish off the call. So this is still within the DoGet RPC even if control flow is outside the Cython do_get code.